### PR TITLE
algod: Simulate Endpoint

### PR DESCRIPTION
**Pull request by algochoi**
This PR:
* Generates Simulate-related objects/types
* Adds algod support for the simulate endpoint
* Implements cucumber tests from this [SDK testing PR](https://github.com/algorand/algorand-sdk-testing/pull/265)

To support sending unsigned transactions to the simulate endpoint, also adds `encodeNoSigTransaction()` to encode them. Currently, algod does not seem to decode `Transaction`, so we wrap them in a `SignedTransaction`-like object, but set the sig field to null. For simulating unsigned txns in the ATC, added `makeNoSigTransactionSigner` to allow the transaction signer to encode unsigned transactions and simulate them.

TODO:
* SDK Tests need to be merged in first
* Need to fix CI tests due to some issues with indexer in dev mode (local tests pass for the new `@simulate` tag)

### CODE REVIEW

File: .test-env [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # Configs for testing repo download:
 SDK_TESTING_URL="https://github.com/algorand/algorand-sdk-testing"
-SDK_TESTING_BRANCH="master"
+SDK_TESTING_BRANCH="simulate-endpoint-tests"
```
Comment (algochoi) [inline .test-env:3]: TODO: revert back to master

File: src/client/v2/algod/algod.ts [hunk 2]
```diff
@@ -575,4 +576,22 @@ export default class AlgodClient extends ServiceClient {
...
+
+  /**
+   * Simulate a list of a signed transaction objects being sent to the network.
+   *
+   * #### Example
+   * ```typescript
+   * const resp = await client.simulateRawTransactions([signedTxn1, signedTxn2]).do();
+   * ```
+   *
+   * [Response data schema details](https://developer.algorand.org/docs/rest-apis/algod/v2/#post-v2transactionssimulate)
```
Comment (algochoi) [inline src/client/v2/algod/algod.ts:588]: Double check this

Comment (algochoi) [reply src/client/v2/algod/algod.ts:588]: confirmed

**Review by tzaffi — COMMENTED** [review]
Overall, looks good. I had several possible nits and other very minor suggestions.

File: src/client/v2/algod/models/types.ts [hunk 3]
```diff
@@ -3084,6 +3084,228 @@ export class PostTransactionsResponse extends BaseModel {
...
+ */
+export class SimulateResponse extends BaseModel {
+  /**
+   * The round immediately preceding this simulation. State changes through this
+   * round were used to run this simulation.
+   */
+  public lastRound: number | bigint;
+
+  /**
+   * A result object for each transaction group that was simulated.
```
[the remark this answers is not in the document]
Comment (algochoi) [reply src/client/v2/algod/models/types.ts:3098]: The `types.ts` files are auto-generated, so we would need to change the upstream generator file/algod API spec.

File: src/client/v2/algod/models/types.ts [hunk 4]
```diff
@@ -3084,6 +3084,228 @@ export class PostTransactionsResponse extends BaseModel {
...
+  // eslint-disable-next-line camelcase
+  static from_obj_for_encoding(data: Record<string, any>): SimulateResponse {
+    /* eslint-disable dot-notation */
+    if (typeof data['last-round'] === 'undefined')
+      throw new Error(
+        `Response is missing required field 'last-round': ${data}`
+      );
+    if (!Array.isArray(data['txn-groups']))
+      throw new Error(
+        `Response is missing required array field 'txn-groups': ${data}`
```
[the remark this answers is not in the document]
Comment (algochoi) [reply src/client/v2/algod/models/types.ts:3158]: This is an autogenerated file, so a manual change would be reverted back eventually unless we change it upstream

**Review by tzaffi — APPROVED** [review]
LGTM

File: src/composer.ts [hunk 5]
```diff
@@ -698,4 +736,46 @@ export class AtomicTransactionComposer {
...
+      if (method.returns.type !== 'void') {
+        const logs: string[] = pendingInfo.logs || [];
+        if (logs.length === 0) {
+          throw new Error('App call transaction did not log a return value');
+        }
+
+        const lastLog = Buffer.from(logs[logs.length - 1], 'base64');
+        if (
+          lastLog.byteLength < 4 ||
+          !lastLog.slice(0, 4).equals(RETURN_PREFIX)
```
Comment (bbroder-algo) [inline src/composer.ts:765]: why do we name return_prefix but assume its length is 4

Comment (algochoi) [reply src/composer.ts:765]: [ARC-04](https://github.com/algorandfoundation/ARCs/blob/main/ARCs/arc-0004.md#standard-format) magic number - the [change predates this PR](https://github.com/algorand/js-algorand-sdk/pull/743/files#diff-534b9601ad1de4bb83d07dc0cb5af85424ccf602bcd8f4c0a2d3b79def06e8aeL677), but seems reasonable to specify a constant here or take a byteLength

**Pull request merged**